[libcxx][test] Do not assume array::iterator is a pointer

strega-nil opened the pull request:
In the tests I added for `ranges::find_last{_if{_not}}`, I accidentally introduced an assumption that `same_as<array<T, 0>::iterator, T*>`; this is a faulty assumption on MSVC-STL.

Fixes #100498.

cc @StephanTLavavej, can you test this PR with the MSVC-STL?

strega-nil committed:
[libcxx] Do not assume array::iterator is a pointer

In the tests I added for `ranges::find_last{_if{_not}}`, I accidentally introduced an assumption
that `same_as<array<T, 0>::iterator, T*>`; this is a faulty assumption on MSVC-STL.

strega-nil committed:
make sure to include <memory>

ldionne reviewed (approved):
This LGTM once CI is green, but I am surprised that this wasn't caught by our unstable-ABI job. That job should define `_LIBCPP_ABI_USE_WRAP_ITER_IN_STD_ARRAY` and `_LIBCPP_ABI_USE_WRAP_ITER_IN_STD_STRING_VIEW`, which use `__wrap_iter` in `array` and `string_view` and should catch those issues.

StephanTLavavej commented:
@strega-nil There are still errors with MSVC's STL. Full logs: [clang_errors.txt](https://github.com/user-attachments/files/16384329/clang_errors.txt)

Here's my analysis of the first error, which is:

```
D:\GitHub\STL\llvm-project\libcxx\test\std\algorithms\alg.nonmodifying\alg.find.last\ranges.find_last_if.pass.cpp(104,21): error: cannot cast from type 'iterator' (aka '_Array_iterator<int, 0>') to pointer type 'it' (aka 'const std::_Array_iterator<int, 0> *')
  104 |       assert(ret == it(a.begin()));
      |                     ^~~~~~~~~~~~
C:\Program Files (x86)\Windows Kits\10\include\10.0.22621.0\ucrt\assert.h(40,17): note: expanded from macro 'assert'
   40 |             (!!(expression)) ||                                                              \
      |                 ^~~~~~~~~~
D:\GitHub\STL\llvm-project\libcxx\test\std\algorithms\alg.nonmodifying\alg.find.last\ranges.find_last_if.pass.cpp(193,3): note: in instantiation of function template specialization 'test_iterator_classes<add_const_to_ptr_t, std::type_identity_t>' requested here
  193 |   test_iterator_classes<add_const_to_ptr_t, std::type_identity_t>();
      |   ^
```

https://github.com/strega-nil/llvm-project/blob/04a67b5617d11a94627c5ee89f24e2fff9ce8eda/libcxx/test/std/algorithms/alg.nonmodifying/alg.find.last/ranges.find_last_if.pass.cpp#L104

https://github.com/strega-nil/llvm-project/blob/04a67b5617d11a94627c5ee89f24e2fff9ce8eda/libcxx/test/std/algorithms/alg.nonmodifying/alg.find.last/ranges.find_last_if.pass.cpp#L188-L193

It appears that `add_const_to_ptr_t` contains the assumption, since it's using `remove_pointer_t` on an array iterator.

strega-nil commented:
> This LGTM once CI is green, but I am surprised that this wasn't caught by our unstable-ABI job. That job should define `_LIBCPP_ABI_USE_WRAP_ITER_IN_STD_ARRAY` and `_LIBCPP_ABI_USE_WRAP_ITER_IN_STD_STRING_VIEW`, which use `__wrap_iter` in `array` and `string_view` and should catch those issues.

It looks like defining `_LIBCPP_ABI_USE_WRAP_ITER_IN_STD_ARRAY` is not enough to get `array<T, 0>::iterator` to be `int*`...

strega-nil committed:
fix array<T, 0>::iterator under ABI_USE_WRAP_ITER

additionally, actually fix the rest of the tests now that I can actually
test them.

strega-nil commented:
@ldionne would you mind taking a look again? I've now actually changed product code (in `array`), since it failed to wrap the iterator for `array<T, 0>`.

StephanTLavavej commented:
The tests are now passing for MSVC's STL (completely with Clang; with MSVC modulo a compiler bug in `constexpr` evaluation). Thanks!

mordante commented on libcxx/include/array:
```diff
@@ -270,18 +270,23 @@ struct _LIBCPP_TEMPLATE_VIS array {
+  using size_type              = size_t;
+  using difference_type        = ptrdiff_t;
+  using reverse_iterator       = std::reverse_iterator<iterator>;
+  using const_reverse_iterator = std::reverse_iterator<const_iterator>;
```
Nice, thanks!

mordante commented on libcxx/test/support/test_iterators.h:
```diff
@@ -339,59 +339,91 @@ cpp20_random_access_iterator(It) -> cpp20_random_access_iterator<It>;
     typedef It                                                 pointer;
     typedef typename std::iterator_traits<It>::reference       reference;
-    typedef typename std::remove_pointer<It>::type             element_type;
+    typedef value_type element_type;
```
please fix the alignment with the lines above.

mordante commented on libcxx/test/support/test_iterators.h:
```diff
@@ -339,59 +339,91 @@ cpp20_random_access_iterator(It) -> cpp20_random_access_iterator<It>;
+      it_ -= n;
+      return *this;
+    }
+    friend constexpr contiguous_iterator operator+(contiguous_iterator x, difference_type n) {
```
Why change this? It at least makes the class not usable in C++03. This function it can't be constexpr in C++11, it's more than just a return statement.

strega-nil replied:
This is in a C++17 block.

mordante reviewed (commented):
In general it looks good, but I'd like to understand the `constexpr` changes for the `contiguous_iterator`

philnik777 commented on libcxx/include/array:
```diff
@@ -270,18 +270,23 @@ struct _LIBCPP_TEMPLATE_VIS array {
+  using const_reference = const value_type&;
+  using pointer         = value_type*;
+  using const_pointer   = const value_type*;
+#if defined(_LIBCPP_ABI_USE_WRAP_ITER_IN_STD_ARRAY)
+  using iterator       = __wrap_iter<pointer>;
+  using const_iterator = __wrap_iter<const_pointer>;
+#else
+  using iterator       = pointer;
+  using const_iterator = const_pointer;
+#endif
```
Can we put this in a separate patch? Technically this could be ABI breaking, since we change the meaning of `_LIBCPP_ABI_USE_WRAP_ITER_IN_STD_ARRAY`. I'm pretty sure it's fine in this case, since we've only introduced this macro recently, but this is potentially API breaking for people (see this very patch). It's also not dependent on the rest of this patch or vice versa.

strega-nil replied:
Sure, I can do that.

strega-nil committed:
CRs - remove array changes

strega-nil committed:
format

Zingam commented:
@strega-nil Off-topic question are the changes to array and contigious_iterator supposed to fix this type of an error:
>  error: no matching conversion for functional-style cast from 'std::__wrap_iter<int *>' to 'contiguous_iterator<int *>'

strega-nil commented:
> @strega-nil Off-topic question: are the changes to array and contigious_iterator supposed to fix this type of an error:
> 
> > error: no matching conversion for functional-style cast from 'std::__wrap_iter<int *>' to 'contiguous_iterator<int *>'

No, these would fix `contiguous_iterator<std::__wrap_iter<int*>>` to work at all.

cjdb commented on libcxx/test/std/algorithms/alg.nonmodifying/alg.find.last/ranges.find_last_if.pass.cpp:
```diff
@@ -183,8 +183,17 @@ struct NonConstComparable {
   friend constexpr bool operator==(NonConstComparable&, const NonConstComparable&) { return true; }
 };
 
+// note: this should really use `std::const_iterator`
```
Can you open an issue and change this to a TODO instead please? AIUI we're only not doing that because `const_iterator` isn't around yet.

strega-nil replied:
Sure.

cjdb replied:
Oh, I meant using the issue in the todo as `TODO(<#issue-number>): <reason>`.

cjdb replied:
> Oh, I meant using the issue in the todo as `TODO(<#issue-number>): <reason>`.

This was a part of my conditional LGTM. I will make it clearer in future reviews.

cjdb reviewed (approved):
LGTM with one request!

strega-nil committed:
switch a note to a TODO

strega-nil commented:
Merging as the only failure was agent lost, and the previous commit (before a comment-only change) did pass.
